presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationMetadata.java

```diff
@@ -146,23 +140,28 @@ private static void verifyInputFunctionSignature(MethodHandle method, List<Param
                     throw new IllegalArgumentException("Unsupported parameter: " + metadata.getParameterType());
             }
         }
+        checkArgument(stateIndex == stateInfos.size(), String.format("Input function only has %d states, expected: %d.", stateIndex, stateInfos.size()));
     }
 
-    private static void verifyCombineFunction(MethodHandle method, Class<?> stateInterface)
+    private static void verifyCombineFunction(MethodHandle method, List<AccumulatorStateInfo> stateInfos)
     {
         Class<?>[] parameterTypes = method.type().parameterArray();
-        checkArgument(parameterTypes.length == 2, "Combine function must take exactly 2 arguments.");
-        checkArgument(Arrays.stream(parameterTypes).filter(type -> type.equals(stateInterface)).count() == 2, "Combine function must take exactly two arguments of type %s annotated as @AggregationState", stateInterface.getSimpleName());
+        checkArgument(parameterTypes.length == stateInfos.size() * 2, "Number of arguments for combine function must be 2 times the size of states.");
+        for (int i = 0; i < stateInfos.size() * 2; i++) {
+            checkArgument(parameterTypes[i].equals(stateInfos.get(i % stateInfos.size()).getStateInterface()), String.format("Type for Parameter index %d is unexpected", i));
+        }
     }
 
-    private static void verifyExactOutputFunction(MethodHandle method, Class<?> stateInterface)
+    private static void verifyExactOutputFunction(MethodHandle method, List<AccumulatorStateInfo> stateInfos)
     {
         if (method == null) {
             return;
         }
         Class<?>[] parameterTypes = method.type().parameterArray();
-        checkArgument(parameterTypes.length == 2, "Output function must take at exactly 2 arguments.");
-        checkArgument(Arrays.stream(parameterTypes).filter(type -> type.equals(stateInterface)).count() == 1, "Output function must take exactly one @AggregationState of type %s", stateInterface.getSimpleName());
+        checkArgument(parameterTypes.length == stateInfos.size() + 1, "Number of arguments for combine function must be one more than number of states.");
```

Comment:
Number of arguments for combine function must be exactly one plus than number of states.